source/index.ts

```diff
@@ -109,49 +104,49 @@ function is(value: any): TypeName { // tslint:disable-line:only-arrow-functions
+	export const nullOrUndefined = (value: null | undefined | any): value is null | undefined => null_(value) || undefined(value);
+	export const object = (value: object) => !nullOrUndefined(value) && (function_(value) || isObject(value));
+	export const iterable = <T = any>(value: IterableIterator<T>): value is IterableIterator<T> => !nullOrUndefined(value) && function_(value[Symbol.iterator]);
+	export const generator = (value: Generator): value is Generator => iterable(value as IterableIterator<any>) && function_(value.next) && function_(value.throw);
 
-	export const nativePromise = isObjectOfType<Promise<any>>(TypeName.Promise);
+	export const nativePromise = <T = any>(value: Promise<T>): value is Promise<T> =>
+		isObjectOfType<Promise<T>>(TypeName.Promise)(value);
 
-	const hasPromiseAPI = (value: any): value is Promise<any> =>
+	const hasPromiseAPI = <T = any>(value: Promise<T>): value is Promise<T> =>
 		!null_(value) &&
 		isObject(value) &&
 		function_(value.then) &&
 		function_(value.catch);
 
-	export const promise = (value: any): value is Promise<any> => nativePromise(value) || hasPromiseAPI(value);
+	export const promise = <T = any>(value: Promise<T>): value is Promise<T> => nativePromise<T>(value) || hasPromiseAPI<T>(value);
 
-	const isFunctionOfType = <T>(type: TypeName) => isObjectOfType<T>(type);
-	export const generatorFunction = isFunctionOfType<GeneratorFunction>(TypeName.GeneratorFunction);
-	export const asyncFunction = isFunctionOfType<Function>(TypeName.AsyncFunction);
-	export const boundFunction = (value: any): value is Function => function_(value) && !value.hasOwnProperty('prototype');
+	export const generatorFunction = isObjectOfType<GeneratorFunction>(TypeName.GeneratorFunction);
+	export const asyncFunction = isObjectOfType<Function>(TypeName.AsyncFunction);
+	export const boundFunction = (value: Function): value is Function => function_(value) && !value.hasOwnProperty('prototype');
 
 	export const regExp = isObjectOfType<RegExp>(TypeName.RegExp);
 	export const date = isObjectOfType<Date>(TypeName.Date);
 	export const error = isObjectOfType<Error>(TypeName.Error);
-	export const map = isObjectOfType<Map<any, any>>(TypeName.Map);
-	export const set = isObjectOfType<Set<any>>(TypeName.Set);
-	export const weakMap = isObjectOfType<WeakMap<any, any>>(TypeName.WeakMap);
-	export const weakSet = isObjectOfType<WeakSet<any>>(TypeName.WeakSet);
+	export const map = <T1 = any, T2 = any>(value: Map<T1, T2>): value is Map<T1, T2> => isObjectOfType<Map<T1, T2>>(TypeName.Map)(value);
+	export const set = <T = any>(value: Set<T>): value is Set<T> => isObjectOfType<Set<T>>(TypeName.Set)(value);
+	export const weakMap = <T1 extends object = any, T2 = any>(value: WeakMap<T1, T2>): value is WeakMap<T1, T2> => isObjectOfType<WeakMap<T1, T2>>(TypeName.WeakMap)(value);
+	export const weakSet = <T extends object = any>(value: WeakSet<T>): value is WeakSet<T> => isObjectOfType<WeakSet<T>>(TypeName.WeakSet)(value);
```

Comment:
All of these `map`, `set`, `weakMap`, `weakSet` should not use generics. See above for details.